sys/linux/socket.txt

```diff
@@ -317,6 +321,41 @@ ifr_ifru [
 
 define IFREQ_SIZE	sizeof(struct ifreq)
 
+# https://elixir.bootlin.com/linux/v5.3.11/source/include/uapi/linux/if_vlan.h
+vlan_args {
+	cmd		flags[vlan_cmd, intptr]
+	device1		array[int8, 24]
+	u		vlan_u
+	vlan_qos	int8
+}
+
+vlan_u [
+	device2		array[int8, 24]
+	VID		intptr
```

Comment:
s/intptr/int32/
Also, will be better to do: int32[0:4] as we already use these small vid's in vnet descriptions:
```
sys/linux/vnet.txt:	vid	int16:12[0:4]
sys/linux/vnet.txt:	vid	int16:12[0:4]
```

Reply:
Updated, and may I know the reason behind that? I didn't see there is a valid range for vid, is that just narrow down the scope of random value and make fuzzer more efficient as well?
Also as I am not able to find where are those struct in vnet.txt are defined in kernel. I am not sure what does the
int16:12[0:4]
means here, the format int16:12 means a random 12 bit bitfield, but how's that compact with [0:4] which is 3 bit wide? Like is that going to randomly generate 12 bit in 4 group and each off them are 0 to 4?

Reply:
This is still confusing me, do you know what's the actually meaning of int16:12[0:4]?

Reply:
The meaning is described here:
https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions_syntax.md#ints
If something there is unclear, what exactly? we then need to improve docs.

The reason why this is done is, yes, to narrow down the values and make matches more likely. Say, if you take network ports, there are 64K of them. But testing exactly ports 34567 vs 34568 does not make lots of sense, but on the other hand, issuing 2 syscalls/packets that operate on the same port makes lots of sense for testing. Or if you take filename, there is infinite space of possible file names, but enumerating all of them makes little sense, but again operating on the same file makes lots of sense. The same relates to vnet vid's.

Reply:
I think I do have read that document carefully, however the point is I am able to get the meaning of int16[0:4], I also get the meaning of int16:12. But that int16:12[0:4] still makes me confused. Checking some fuzz input, it looks like the value is just 0 to 4. If it is going to be value from 0 to 4, why we need that 12 bitfield(3 is enough)? Border question is why do we need bitfield while the range of value is given? Is that something to do with the endian? Sorry if this is a stupid question..

Reply:
Bitfield part (:12) determines physical layout of data in memory, while ([0:4]) determines what values we pass there. These are completely orthogonal things. Consider:
```
X {
    a int64[0:2]
    b int8
}
```
Here we pass just 3 values in a, but kernel will read field b from 9-th byte, thus we need to say that a takes first 7 bytes in memory. If we do:
```
X {
    a int8[0:2]
    b int8
}
```
We still can perfectly pass 4 values in a, _but_ now b starts at 2-nd byte and if kernel expects it at 9-th bytes, we messed the layout -- we will be passing b not where kernel expects it.
Bitfields control physical layout, consider these two are laid out in memory totally differently:
```
X {
    a int32:4
    b int32:6
    c int32:3
}
```
```
X {
    a int32
    b int32
    c int32
}
```

Reply:
Thanks a lot! That makes lot of sense to me now. Sorry I was massed up with the bitfield earlier, I was thought like int32:4 means we will only randomly assign last 4 bit while while 32 bit is being used. Which is actually totally wrong. The document did indicate that point, but just when I am reading I wasn't thinking in that way..